### PR TITLE
Result object

### PR DESCRIPTION
**Pull request by Framba-Luca**
Implemented the FinalReport and IntermediateReport class with all the needed adjustments and new parameter to the solve method to send the IntermediateReports.

This solves #76

### CODE REVIEW

File: unified_planning/plan.py
Comment (alvalentini) [inline unified_planning/plan.py:118]: I think it is a good idea to refactor the common fields in a base class

File: unified_planning/solvers/parallel.py
Comment (alvalentini) [inline unified_planning/solvers/parallel.py:94]: you are right! Here we should handle all possible plans

File: unified_planning/plan.py
Comment (alvalentini) [inline unified_planning/plan.py:33]: Should we include also TIMEOUT, MEMOUT, INTERNAL_ERROR?

**Review by alvalentini — CHANGES_REQUESTED** [review]

**Comment by alvalentini** [comment]
The `IntermediateReport` and the `FinalReport` proposed by @arbimo in the proto file share the same contents except for the `Status` field. @Framba-Luca followed the same structured in the UP implementation of this PR.

I'm wondering if it would be better to have a `Answer` message/class that has the same fields as the `FinalReport` with an additional possible value `INTERMEDIATE` for the `Status`.

```
message Answer {
    enum Status {
        // ==== Planner stopped normally ======

        // Valid plan found and search stopped immediately
        // The `best_plan` field must be set.
        SAT = 0;
        // Plan found with optimality guarantee
        // The `best_plan` field must be set and contain an optimal solution.
        OPT = 1;
        // No plan exists
        UNSAT = 2;
        // The planner was not able to find a solution but does not give any guarantee that none exist
        // (i.e. the planner might not be complete)
        SEARCH_SPACE_EXHAUSTED = 3;

        // ====== Planner exited before making any conclusion ====
        // Search stopped before concluding OPT or UNSAT
        // If a plan was found, it might be reported in the `best_plan` field

        // The planner ran out of time
        TIMEOUT = 13;
        // The planner ran out of memory
        MEMOUT = 14;
        // The planner faced an internal error.
        INTERNAL_ERROR = 15;
        // The problem submitted is not supported by the planner.
        UNSUPPORTED_PROBLEM = 16;

        // ====== Intermediate answer ====
        INTERMEDIATE = 17;
    }
    Status status = 1;

    // Optional. Best plan found so far, if any.
    Plan best_plan = 2;

    // A set of planner specific values that can be reported, for instance
    // - "grounding-time": "10ms"
    // - "expanded-states": "1290"
    map<string, string> metrics = 3;

    // Optional logs about the planner's activity.
    // Note that it should not be expected that logging messages are visible to the end user.
    // If used in conjunction with INTERNAL_ERROR or UNSUPPORTED_PROBLEM, it would be expected to have at least one log message at the ERROR level.
    repeated LogMessage logs = 4;
}
```

@arbimo @mikand what do you think?

**Comment by arbimo** [comment]
@alvalentini Yes, I think it makes sense to merge them into a single one. While they were initially clearly distinct, they essentially converged over time. 
It might make sense to keep the two classes in the python API however (just like for actions).

In your proposal it might make sense to rename `best_plan` to just `plan` which is less awkward to use. With a comment specifying that this is a plan that should not be dominated by a previous plan and that it should be left empty if it is an intermediate message and no plan has been found since the previous reported one.

File: unified_planning/plan.py
Comment (Framba-Luca) [reply unified_planning/plan.py:33]: I don't think so because NEGATIVE_OUTCOMES assumes that a plan is not found. TIMEOUT, MEMOUT and INTERNAL_ERROR, at least from my interpretation of their meaning, could return a plan, still knowing that something went wrong.

Comment (alvalentini) [reply unified_planning/plan.py:33]: I don't agree. They are reasons why a plan is not found.
TIMEOUT for example means that a plan is not found within the time limit

Comment (arbimo) [reply unified_planning/plan.py]: In the case of planning with explicit costs, it could be the case that a plan was found but the planner exited (e.g. timeout) before finding the optimal one. In this case, the "best plan found" remains very important.

Comment (alvalentini) [reply unified_planning/plan.py]: Do you think this is the case also for MEMOUT, INTERNAL_ERROR and SEARCH_SPACE_EXHAUSTED?

Comment (alvalentini) [reply unified_planning/plan.py]: From what I understood, SATISFIED and OPTIMAL requires the `plan` field set while the UNSATISFIED and UNSUPPORTED_PROBLEM never have the `plan` field set. In the other cases a plan might be set, right?

File: unified_planning/plan.py
Comment (alvalentini) [inline unified_planning/plan.py:206]: Typo:
```suggestion
    SATISFIED: 'SATISFIED',
```

Check to use the same name here and in the proto file.

Comment (Framba-Luca) [reply unified_planning/plan.py:206]: Should we use SAT or SATISFIED? I personally prefer the long option (more clear), but it's not really something major.

File: unified_planning/plan.py
Comment (arbimo) [reply unified_planning/plan.py]: Correct except for SEARCh_SPACE_EXHAUSTED which that the planner stopped because he is not capable of generating more potential solutions (but there might other that are possible, i.e., the planner is incomplete). 

In a nutshell :
-  if a plan is provided with the SEARCH_SPACE_EXHAUSTED flag, it is a weaker version of the OPTIMAL flag (without an actual guarantee on the optimality).
- if no plan is provided with the SEARCH_SPACE_EXHAUSTED flag, it is a weaker version of the UNSATISFIABLE version : the planner reports that he is not capable of finding a solution but a complete planner might find one.

I was unsure whether to include this flag. The reasoning was that (1) we want to trust the planner when it says that something is optimal/unsat because this is not something we can routinely verify, (2) if we don't include those weaker cases, then incomplete planners will have to (incorrectly) report optimal/unsat.

This is mostly related to planner incompletness so it might handle another way (as distinct boolean flag, directly in the planner description, ...)

Comment (alvalentini) [reply unified_planning/plan.py]: I agree on what you said about the case "no plan is provided with the SEARCH_SPACE_EXHAUSTED flag".
For the case "a plan is provided with the SEARCH_SPACE_EXHAUSTED flag", I think we can express the same message with the SATISFIED flag: the planner reports that the plan is the best one found so far (with no guarantee on the optimality) and it is not able to find any better plans.

**Review by alvalentini — CHANGES_REQUESTED** [review]
Except for the minor comment, I think the PR is ready to merge.

**Review by alvalentini — APPROVED** [review]

**Review by mikand — APPROVED** [review]
@Framba-Luca good work! I left some minor comments, but this looks OK.
@roeger @hstairs @LBonassi95: let us know if it is OK to merge this or if you have any feedback.

File: unified_planning/solvers/pddl_solver.py
Comment (roeger) [reply unified_planning/solvers/pddl_solver.py]: No problem from my side.

**Review by roeger — COMMENTED** [review]
The changes look good. I am only unsure about the terminology "satisfied/unsatisfied". What is satisfied or unsatisfied here? From my understanding "solved" and "unsolvable" would be the idiomatic terms for what it should express according to the comments. In particular, I don't like satisfied so much, because my brain always wants to connect it to satisficing planning (which obviously is not called satisfying planning).

**Comment by mikand** [comment]
@arbimo @roeger To be clear, here we have three "successful outcomes":
- SATISFIED/SAT: A valid plan has been found, the plan is not necessarily optimal 
- UNSOLVABLE/UNSAT: We proved that no valid plan exists
- OPTIMAL: An optimal plan has been found (proving that no better plan can exist)

Now, for the naming I like SATISFIED/UNSOLVABLE/OPTIMAL better than SAT/UNSAT/OPTIMAL, but this is a matter of taste. 

Do you agree on the three cases above? Any feeedback/suggestion?

**Comment by arbimo** [comment]
> Now, for the naming I like SATISFIED/UNSOLVABLE/OPTIMAL better than SAT/UNSAT/OPTIMAL, but this is a matter of taste.
@mikand Yes, I think it is a perfectly fine combination :+1:   (my personal preference would have been satisfied/unsatisfiable/optimal but as you said it is a matter of taste and "unsolvable" might more common in the planning community).

**Comment by roeger** [comment]
I also don't like the asymmetry between SOLVED (or SATISFIED) and OPTIMAL because the latter implies the former. If they were not so long, I would argue for SOLVED_WITHOUT_OPTIMALITY_GUARANTEE / DETECTED_UNSOLVABLE / SOLVED_WITH_OPTIMALITY_GUARANTEE.

**Comment by hstairs** [comment]
we could have SOLVED_OPTIMALLY/SOLVED/UNSOLVABLE. Or we can leave that to be SOLVED/UNSOLVED and have an OPTIMALITY_GUARANTEE field to be true or false. This last option is what I like the most because it is more factored; not sure though where that will fit in the Answer message defined above, probably outside "status"?

File: unified_planning/solvers/parallel.py
Comment (alvalentini) [inline unified_planning/solvers/parallel.py:98]: Here I would keep only UNSOLVABLE_INCOMPLETELY

Comment (Framba-Luca) [reply unified_planning/solvers/parallel.py:98]: Why? If a planner went in TIMEOUT it would make sense to let the user know that with more time a plan might be found, while UNSOLVABLE_INCOMPLETELY means that, even with more time, those planners will not find a plan.

File: unified_planning/solvers/parallel.py
Comment (Framba-Luca) [reply unified_planning/solvers/parallel.py:90]: Why not?

Comment (alvalentini) [reply unified_planning/solvers/parallel.py:90]: Because when the status is UNSOLVABLE_INCOMPLETELY, no plans are present in the result

Comment (Framba-Luca) [reply unified_planning/solvers/parallel.py:90]: True.

File: unified_planning/solvers/parallel.py
Comment (alvalentini) [inline unified_planning/solvers/parallel.py:94]: Here you have at least to connect the status with its planner name.

File: unified_planning/solvers/parallel.py
Comment (alvalentini) [reply unified_planning/solvers/parallel.py:100]: This is not solved. You need to remove the timeout.

**Review by alvalentini — CHANGES_REQUESTED** [review]

File: unified_planning/solvers/parallel.py
Comment (mikand) [inline unified_planning/solvers/parallel.py:75]: I would terminate also when a SATISFICING plan is returned if the problem (args[0]) has no optimality metrics.

**Review by mikand — APPROVED** [review]
I left two minor comments, after those, @alvalentini  please go ahead and merge

**Pull request merged**
